src/security/body_parse/body_parsing.cpp

```diff
@@ -0,0 +1,167 @@
+#include "body_parsing.h"
+
+#include <ddwaf.h>
+
+#include <cstddef>
+#include <unordered_map>
+
+#include "../ddwaf_memres.h"
+#include "../ddwaf_obj.h"
+#include "../decode.h"
+#include "body_json.h"
+#include "body_multipart.h"
+#include "chain_is.h"
+#include "header.h"
+
+extern "C" {
+#include <ngx_core.h>
+}
+
+using namespace std::literals;
+
+namespace dnsec = datadog::nginx::security;
+
+namespace {
+
+bool is_content_type(std::string_view actual, std::string_view tested) {
+  auto sv = actual;
+  while (sv.front() == ' ' || sv.front() == '\t') {
+    sv.remove_prefix(1);
+  }
+
+  if (sv.size() < tested.size()) {
+    return false;
+  }
+
+  for (std::size_t i = 0; i < tested.size(); i++) {
+    if (std::tolower(sv.at(i)) != tested.at(i)) {
+      return false;
+    }
+  }
+  sv.remove_prefix(tested.size());
+
+  return sv.empty() || sv.front() == ';' || sv.front() == ' ' ||
+         sv.front() == '\t';
+}
+
+bool is_json(const ngx_http_request_t &req) {
+  const ngx_table_elt_t *ct = req.headers_in.content_type;
+  // don't look at ct->next; consider only the first value
+  return ct && is_content_type(datadog::nginx::to_string_view(ct->value),
+                               "application/json"sv);
+}
+
+bool is_multipart(const ngx_http_request_t &req) {
+  const ngx_table_elt_t *ct = req.headers_in.content_type;
+  return ct && is_content_type(datadog::nginx::to_string_view(ct->value),
+                               "multipart/form-data"sv);
+}
+
+bool is_urlencoded(const ngx_http_request_t &req) {
+  const ngx_table_elt_t *ct = req.headers_in.content_type;
+  return ct && is_content_type(datadog::nginx::to_string_view(ct->value),
+                               "application/x-www-form-urlencoded"sv);
+}
+bool is_text_plain(const ngx_http_request_t &req) {
+  const ngx_table_elt_t *ct = req.headers_in.content_type;
+  return ct && is_content_type(datadog::nginx::to_string_view(ct->value),
+                               "text/plain"sv);
+}
+
+}  // namespace
+
+namespace datadog::nginx::security {
+
+bool parse_body(ddwaf_obj &slot, ngx_http_request_t &req,
+                const ngx_chain_t &chain, std::size_t size,
+                DdwafMemres &memres) {
+  if (is_json(req)) {
+    // use rapidjson to parse:
+    bool success = parse_json(slot, req, chain, memres);
+    if (success) {
+      return true;
+    }
+  } else if (is_multipart(req)) {
+    std::optional<HttpContentType> ct = HttpContentType::for_string(
+        to_string_view(req.headers_in.content_type->value));
+    if (!ct) {
+      ngx_log_error(NGX_LOG_NOTICE, req.connection->log, 0,
+                    "multipart: invalid multipart/form-data content-type");
+      return false;
+    }
+
+    return parse_multipart(slot, req, *ct, chain, memres);
+  }
+
+  bool ct_plain = is_text_plain(req);
+  bool ct_urlencoded = !ct_plain && is_urlencoded(req);
+
+  if (!ct_plain && !ct_urlencoded) {
```

Comment:
the content-type cannot be plain and urlencoded at the same time; logically `bool ct_urlencoded = !ct_plain && is_urlencoded(req);` is equivalent to the simpler `bool ct_urlencoded = is_urlencoded(req);`. However, the version with `!ct_plain` avoids doing the comparison if we already now it's text/plain. It's basically a (marginal) optimization.

Reply:
> the content-type cannot be plain and urlencoded at the same time;

That being said, may I ask, why not just `bool ct_urlencoded = is_urlencoded(req);` then? That's details but I am not sure there is any optimization here.

Reply:
Alright, just for the record, because I was looking for the optimization somewhere else, the optimization is not calling `is_urlencoded(req)` if it's plain text.